### PR TITLE
Full support for websocket reconnection/resubscription

### PR DESCRIPTION
**Pull request by cgewecke**
This is #3135, #1966 redux, merged to the root repo so merge conflicts and tests can be managed here. Tests are WIP, opening as draft.

### CODE REVIEW

**Comment by nivida** [comment]
@cgewecke Sorry for commenting on this draft PR. But I would like to ask you what the state is of the mentioned test cases? Is it possible to commit those and I'm gonna finish them?

**Comment by cgewecke** [comment]
@nivida Ah thanks for the reminder - I will get these in..

File: test/eth.subscribe.ganache.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                .on('error', function(err){
+                    assert(err.message.includes("CONNECTION ERROR"))
+                    this.removeAllListeners();
+                    resolve();
+                });
+        });
+    });
+
+    // This test failing....
+    it.only('auto reconnects', function(){
```
Comment (cgewecke) [inline test/eth.subscribe.ganache.js:58]: @nivida This is my attempt to model reconnection working. 
+ Open server
+ Attach a subscription handler
+ Fire a tx (which subscrip hears)
+ Close server
+ Re-open server
+ Fire a tx (which subscrip hears)

But the test doesn't reconnect...do you want to take a look?

Comment (cgewecke) [reply test/eth.subscribe.ganache.js:58]: We need to make sure this works against an actual client, not just mock some assumptions in.

Comment (cgewecke) [reply test/eth.subscribe.ganache.js:58]: Maybe my model of how to do this is wrong...not sure.

Comment (cgewecke) [reply test/eth.subscribe.ganache.js:58]: @nivida mentioned something in a chat that's important which I'm leaving here for further reference. 

The way this test models disconnection doesn't reflect the connectivity issues people typically have online e.g. nodes are not shutting down, killing their sockets and restarting...

**Comment by nivida** [comment]
Moved this to #3190

**Pull request closed without merging**